packages/messages/src/browser/notifications-manager.ts

```diff
@@ -0,0 +1,285 @@
+import { deepClone } from '@theia/core/lib/common/objects';
+import { Event, Emitter } from '@theia/core';
+import { Deferred } from '@theia/core/lib/common/promise-util';
+import { Md5 } from 'ts-md5';
+import * as markdownit from 'markdown-it';
+import throttle = require('lodash.throttle');
+import { NotificationPreferences } from './notification-preferences';
+import { ContextKeyService, ContextKey } from '@theia/core/lib/browser/context-key-service';
+import { OpenerService } from '@theia/core/lib/browser';
+import URI from '@theia/core/lib/common/uri';
+
+export const NotificationManager = Symbol('NotificationManager');
+export interface NotificationManager {
+    readonly open: boolean;
+    readonly onUpdate: Event<NotificationManager.UpdateEvent>;
+    accept(notification: Notification | string, action: string): void;
+    hide(): void;
+    toggle(): void;
+    clear(notification: Notification | string): void;
+    clearAll(): void;
+    toggleExpansion(notification: string): void;
+    openLink(link: string): Promise<void>;
+}
+export namespace NotificationManager {
+    export interface UpdateEvent {
+        readonly notifications: Notification[];
+        readonly open: boolean;
+    }
+}
+
+export interface Notification {
+    messageId: string;
+    location: Notification.Location;
+    message: string;
+    source?: string;
+    expandable: boolean;
+    collapsed: boolean;
+    type: Notification.Type;
+    actions: string[];
+    progress?: number;
+}
+export namespace Notification {
+    export type Type = 'info' | 'warning' | 'error' | 'progress';
+    export type Location = 'window' | 'notification';
+}
+
+@injectable()
+export class NotificationManagerImpl extends MessageClient implements NotificationManager {
+
+    @inject(NotificationPreferences)
+    protected readonly preferences: NotificationPreferences;
+
+    @inject(ContextKeyService)
+    protected readonly contextKeyService: ContextKeyService;
+
+    @inject(OpenerService)
+    protected readonly openerService: OpenerService;
+
+    protected readonly onUpdateEmitter = new Emitter<NotificationManager.UpdateEvent>();
+    protected readonly fireUpdateEvent = throttle(() => {
+        const notifications = deepClone(Array.from(this.notifications.values()));
+        this.onUpdateEmitter.fire({ notifications, open: this.open });
+    }, 250, { leading: false });
+    readonly onUpdate = this.onUpdateEmitter.event;
+
+    protected readonly resultPromises = new Map<string, Deferred<string | undefined>>();
+    protected readonly notifications = new Map<string, Notification>();
+
+    protected notificationToastsVisibleKey: ContextKey<boolean>;
+    protected notificationCenterVisibleKey: ContextKey<boolean>;
+
+    @postConstruct()
+    protected async init(): Promise<void> {
+        this.notificationToastsVisibleKey = this.contextKeyService.createKey<boolean>('notificationToastsVisible', false);
```

Comment:
What about `notificationFocus` context key? It is needed for VS Code compatibility

Reply:
Turns out the focus context requires a separation between toasts and center. I’m revising this part today. Before that I thought it would simplify a lot to not show toasts separately.

Reply:
I'll tackle this separately together with keeb navigation.